### DISCUSSION TITLE
Questions regarding this game engine?

### DISCUSSION THREAD

**Discussion by Raj2032** [17cupsofcoffee/tetra#306] [General]
Hey guys I have some questions regarding this game engine:

1. Does it do multithreading automatically?
2. Does it "lock you" into a framework like [Bevy](https://bevyengine.org/) does or can I code more freely?
3. Will it ever support 3D as well?
4. Will it ever have support for a GUI editor?
5. Does it have its own UI library?

**Comment by 17cupsofcoffee** [comment]
1. Tetra itself does no hidden multithreading.
    * I think the `image` crate (which Tetra uses for image decoding) may use a thread pool for certain things.
    * All of Tetra's rendering *has* to take place on the main thread, but you're free to use other threads in your game logic however you want.
2. The only thing Tetra tries to take control over is the game loop - what you do inside `update` and `draw` is up to you.
3. Probably not - other engines like Bevy and RG3D cover that use case much better. I don't make 3D games, so I don't feel like I have the knowledge neccecary to make a good 3D engine right now.
4. Definitely not - I find general purpose editors massively overkill for 2D games).
    * Other people could build their own editors if they want, though.
5. No, but there's integrations for other GUI libraries out there (e.g. [egui-tetra](https://github.com/tesselode/egui-tetra)).

**Reply by Raj2032** [reply]
OK thanks mate :)
